### PR TITLE
Add generalized harmonic constraint energy

### PR DESCRIPTION
**Pull request by geoffrey4444**
## Proposed changes

Add the generalized harmonic constraint energy (unnormalized). Also update the generalized harmonic constraint documentation to use the new \cite method, instead of just inserting an arXiv URL.

Update: now depends on PR #1246.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: can use `square`

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: the `for` doesn't seem necessary, and the size isn't checked in the case of a `DataVector` for the contained type. The size checking isn't done in the other constraints, so I'll just make an issue to put it in all of them.

**Review by moxcodes — COMMENTED** [review]
Group code review with @markscheel and @roboberlin

**Comment by geoffrey4444** [comment]
Can I squash?

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: Sorry, I've referred you to a piece of code that didn't seem to have implemented all the changes I wanted yet. it is preferable to do this in a single line:
`*energy = Scalar<DataType>{get<0>(f_constraint)};`

**Review by moxcodes — COMMENTED** [review]
One last nitpick, which I'm happy for you to squash immediately along with your current fixup.

**Comment by geoffrey4444** [comment]
Ping @sxs-collaboration/spectre-core-devs 

This needs a second reviewer...could you please assign one or take a look? I think it should be basically ready to go

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (hrueter) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: I wanted to bring this up already in the 4-index constraint PR, but somehow didn't.
What do you think of using the anti-symmetry of the first to indices in this variable?

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (hrueter) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:1264]: I don't quite understand what you are doing here. This is similar to the code from #1172 where we commented
https://github.com/sxs-collaboration/spectre/pull/1172#discussion_r245938351

Is this basically allocating `energy`?

**Review by hrueter — COMMENTED** [review]
Not yet a full review. More is coming next week.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:1264]: There's no guarantee that energy points to something of the correct size. This checks that this size is correct and allocates something of the correct size if not.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: Just set it to the size, don't bother copying the data

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: just don't use `+=`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: Sure...How do you resize without copying the data?

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: Good idea! I've modified the function to take the output of the actual 4-index-constraint function, which is 
a rank-3 tensor containing only the independent components of the 4-index constraint. This lets me also 
eliminate two loops when computing the constraint energy.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:1264]: This is checking that *energy has the correct size and if not, allocating something that does have the correct size.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: Never mind...I'll just use make_with_value

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: I don't understand this comment. How can I do the summations efficiently without +=?

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (nilsdeppe) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: No, don't use `make_with_value`, just do `Scalar<DataType>{get_size(get<0>(f_constraint))}`

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (nilsdeppe) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: You can write the code as:
```cpp
get(*energy) = gauge_constraint_multiplier * square(gauge_constraint.get(0))
               + two_index_constraint_multiplier * square(f_constraint.get(0));
for (size_t a = 1; a < SpatialDim + 1; ++a) {
  get(*energy) += gauge_constraint_multiplier * square(gauge_constraint.get(a))
                  + two_index_constraint_multiplier * square(f_constraint.get(a));
}
// next loop series
```

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: please combine these two `+=` statements into one.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: Seems cumbersome, since I need lots of terms with a=0. Is there a big expected penalty to my way of setting to zero?

Comment (nilsdeppe) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: How is it cumbersome? I just gave you the code you need.

Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: Code seems easier to read to me when you don’t treat a=0 as a special case. That’s what I meant...cumbersome in how it reads, not in difficulty. But if you feel strongly about avoiding the *energy=0 line, of course I can use the code you provided. I just hoped to learn why my approach wasn’t as good, since it seems easier to read and simpler to me.

Comment (nilsdeppe) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: You're doing work that's not necessary every time. I'm really not seeing how the below code is so much more cumbersome to read.

```cpp
if (get_size(get(*energy)) != get_size(get<0>(f_constraint))) {
  *energy = Scalar<DataType>(get_size(get<0>(f_constraint)));
}

get(*energy) = gauge_constraint_multiplier * square(gauge_constraint.get(0)) +
               two_index_constraint_multiplier * square(f_constraint.get(0));
for (size_t a = 1; a < SpatialDim + 1; ++a) {
  get(*energy) +=
      gauge_constraint_multiplier * square(gauge_constraint.get(a)) +
      two_index_constraint_multiplier * square(f_constraint.get(a));
}

for (size_t a = 0; a < SpatialDim + 1; ++a) {
  for (size_t i = 0; i < SpatialDim; ++i) {
    for (size_t j = 0; j < SpatialDim; ++j) {
      get(*energy) +=
          two_index_constraint_multiplier * two_index_constraint.get(i, a) *
          two_index_constraint.get(j, a) * inverse_spatial_metric.get(i, j);

      for (size_t b = 0; b < SpatialDim + 1; ++b) {
        get(*energy) += three_index_constraint_multiplier *
                            three_index_constraint.get(i, a, b) *
                            three_index_constraint.get(j, a, b) *
                            inverse_spatial_metric.get(i, j) +
                        2.0 * four_index_constraint_multiplier *
                            get(spatial_metric_determinant) *
                            four_index_constraint.get(i, a, b) *
                            four_index_constraint.get(j, a, b) *
                            inverse_spatial_metric.get(i, j);
      }
    }
  }
}
```

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: Thanks @nilsdeppe! I didn't realize you could resize like that.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: I feel that splitting the a loop as you do makes it more complicated than keeping the a loop together, but since it sounds like you're saying the extra steps to enable one loop have a likely performance penalty, I guess I don't feel strongly enough about avoiding splitting the loop to insist on it. My latest mixup commit puts this the way you suggested.

**Review by nilsdeppe — DISMISSED** [review]
I'm happy with the changes

**Review by moxcodes — DISMISSED** [review]

**Comment by kidder** [comment]
@hrueter @moxcodes @nilsdeppe please look at

**Comment by geoffrey4444** [comment]
Travis failures are not caused by this PR

1) an unrelated file (`src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.hpp`) was pushed to develop while failing IWYU
2) The usual gcc segfaults

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp]: I'm not sure I like linking to external documents like this. You can convert the latex code to doxygen using `tools/latex2dox.py`.

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: I disagree. Please add a Kerr-Schild test.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
@geoffrey4444 on the latest develop on my machine using the container I get:
```
(/home/nils/SpECTRE/spectre/src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.hpp has correct #includes/fwd-decls)

(/home/nils/SpECTRE/spectre/src/NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.cpp has correct #includes/fwd-decls)
[100%] Built target iwyu
```
I don't see how your claim that something was pushed while failing IWYU is true.

Might have been fixed since then? That’s what the Travis logs said

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp
Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp]: It's a 4-page note with 24 equations. Are you saying you want that much help text here?

Comment (nilsdeppe) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp]: Yep, that's exactly what I'm saying. This was done for Kerr-Schild as well: https://spectre-code.org/classgr_1_1solutions_1_1kerrschild

**Comment by geoffrey4444** [comment]
Rebased on develop, added Kerr test and expanded LaTeX docs. The Kerr test makes this PR depend on #1246.

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: This comment needs to be updated from `four-index constraint` -> `constraint energy`

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp]: this `\emph` isn't showing up correctly. I think the markdown equivalent is `*` around the word. If you found the conversion script useful overall then I can have it automate this change as well in the future.

**Review by nilsdeppe — COMMENTED** [review]
Looks good. Two typo-ish comments for when you squash. I'm happy for you to squash the commits :+1:

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp
Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp]: Thanks @nilsdeppe! I decided in this case the \emph{} isn't needed, so I just removed it. But I would appreciate it if you added \textbf{} and \emph{} to the conversion script. Another helpful thing would be (optionally?) adding the asterisks and spaces on the left hand side.

The conversion script was more than useful: without it, I just wouldn't convert long LaTeX documents into C++/doxygen, because it would be too tedious compared to linking to overleaf, as I did previously. Thanks for making this script!

**Comment by geoffrey4444** [comment]
@nilsdeppe I fixed the typoish points you flagged and squashed. I'll have to rebase on develop after #1246 is merged before this can be merged.

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp
Comment (nilsdeppe) [reply tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: You've added the Kerr-Schild test. When you rebase after the F constraint is merged could you remove that part of this comment?

**Review by nilsdeppe — COMMENTED** [review]
I'm happy with this. One note for when you rebase.

**Comment by nilsdeppe** [comment]
@moxcodes @markscheel @roboberlin Is your group happy with the way your comments were addressed?

**Review by moxcodes — DISMISSED** [review]

**Comment by geoffrey4444** [comment]
Squashed, tests pass. @moxcodes @hrueter @markscheel @roboberlin @nilsdeppe do you approve? Or would you like other changes?

**Review by moxcodes — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**
